src/axes/axis.ts

```diff
@@ -150,9 +168,238 @@ export class Axis<D> extends Component {
                                                     .attr(this._generateTickMarkAttrHash(true));
     tickMarks.exit().remove();
     this._baseline.attr(this._generateBaselineAttrHash());
+    if (this.annotationsEnabled()) {
+      this._drawAnnotations();
+    } else {
+      this._removeAnnotations();
+    }
+    return this;
+  }
+
+  /**
+   * Gets the annotated ticks
+   */
+  public annotatedTicks(): D[];
+  /**
+   * Sets the annotated ticks
+   *
+   * @returns {Axis} The calling Axis.
+   */
+  public annotatedTicks(annotatedTicks: D[]): Axis<D>;
+  public annotatedTicks(annotatedTicks?: D[]): any {
+    if (annotatedTicks == null) {
+      return this._annotatedTicks;
+    }
+    this._annotatedTicks = annotatedTicks;
+    this.render();
+    return this;
+  }
+
+  /**
+   * Gets the formatter for the annotations
+   */
+  public annotationFormatter(): Formatter;
+  /**
+   * Sets the formatter for the annotations
+   *
+   * @returns {Axis} The calling Axis.
+   */
+  public annotationFormatter(annotationFormatter: Formatter): Axis<D>;
+  public annotationFormatter(annotationFormatter?: Formatter): any {
+    if (annotationFormatter == null) {
+      return this._annotationFormatter;
+    }
+    this._annotationFormatter = annotationFormatter;
+    this.render();
+    return this;
+  }
+
+  /**
+   * Gets if annotations are enabled
+   */
+  public annotationsEnabled(): boolean;
+  /**
+   * Sets if annotations are enabled
+   *
+   * @returns {Axis} The calling Axis.
+   */
+  public annotationsEnabled(annotationsEnabled: boolean): Axis<D>;
+  public annotationsEnabled(annotationsEnabled?: boolean): any {
+    if (annotationsEnabled == null) {
+      return this._annotationsEnabled;
+    }
+    this._annotationsEnabled = annotationsEnabled;
+    this.render();
     return this;
   }
 
+  protected _drawAnnotations() {
+    let labelPadding = 4;
+    let measurements = new Utils.Map<D, SVGTypewriter.Measurers.Dimensions>();
+    let annotatedTicks = this._annotatedTicksInDomain();
+    annotatedTicks.forEach((annotatedTick) => {
+      let measurement = this._annotationMeasurer.measure(this.annotationFormatter()(annotatedTick));
+      let paddedMeasurement = { width: measurement.width + 2 * labelPadding, height: measurement.height + 2 * labelPadding };
+      measurements.set(annotatedTick, paddedMeasurement);
+    });
+
+    let tierHeight = this._annotationMeasurer.measure().height + 2 * labelPadding;
+
+    let annotationToTier = this._annotationToTier(measurements);
+
+    let hiddenAnnotations = new Utils.Set<any>();
+    let numTiers = Math.floor(this.margin() / tierHeight);
+    annotationToTier.forEach((tier, annotation) => {
+      if (tier === -1 || tier >= numTiers) {
+        hiddenAnnotations.add(annotation);
+      }
+    });
+
+    let bindElements = (selection: d3.Selection<any>, elementName: string, className: string) => {
```

Comment:
I personally just didn't like looking at the same 3-4 lines of code multiple times that were doing almost the same thing every time.  Also shoring up the code into a single function centralizes the failure points into a single function rather than across multiple implementations of this (I think there were a few times where I accidentally forgot to class the selections or forget to put the `true` in the classed call or use the wrong selector in the `selectAll` etc.)